src/sinks/pulsar.rs

```diff
@@ -0,0 +1,178 @@
+    pending_acks: HashSet<usize>,
+    acker: Acker,
+}
+
+pub type SendFuture =
+    Box<dyn Future<Item = CommandSendReceipt, Error = pulsar::Error> + 'static + Send>;
+
+inventory::submit! {
+    SinkDescription::new_without_default::<PulsarSinkConfig>("pulsar")
+}
+
+#[typetag::serde(name = "pulsar")]
+impl SinkConfig for PulsarSinkConfig {
+    fn build(&self, cx: SinkContext) -> crate::Result<(super::RouterSink, super::Healthcheck)> {
+        let sink = PulsarSink::new(self.clone(), cx.acker(), cx.exec())?;
+        let hc = healthcheck(self, &sink.pulsar)?;
+        Ok((Box::new(sink), hc))
+    }
+
+    fn input_type(&self) -> DataType {
+        DataType::Log
+    }
+
+    fn sink_type(&self) -> &'static str {
+        "pulsar"
+    }
+}
+
+impl PulsarSink {
+    fn new(config: PulsarSinkConfig, acker: Acker, exec: TaskExecutor) -> crate::Result<Self> {
+        let pulsar = Pulsar::new(config.address.parse()?, None, exec).wait()?;
+        let producer = pulsar.producer(Some(ProducerOptions {
+            batch_size: config.batch_size,
+            ..ProducerOptions::default()
+        }));
+
+        Ok(Self {
+            topic: config.topic,
+            encoding: config.encoding,
+            pulsar,
+            producer,
+            in_flight: FuturesUnordered::new(),
+            seq_head: 0,
+            seq_tail: 0,
+            pending_acks: HashSet::new(),
+            acker,
+        })
+    }
+}
+
+impl Sink for PulsarSink {
+    type SinkItem = Event;
+    type SinkError = ();
+
+    fn start_send(&mut self, item: Self::SinkItem) -> StartSend<Self::SinkItem, Self::SinkError> {
+        let message = encode_event(item, &self.topic, self.encoding).map_err(|_| ())?;
+        let fut = self.producer.send(self.topic.clone(), &PulsarSend(message));
+
+        let seqno = self.seq_head;
+        self.seq_head += 1;
+        self.in_flight
+            .push((Box::new(fut) as SendFuture).join(future::ok(seqno)));
+        Ok(AsyncSink::Ready)
+    }
+
+    fn poll_complete(&mut self) -> Poll<(), Self::SinkError> {
+        loop {
+            match self.in_flight.poll() {
+                Ok(Async::NotReady) => return Ok(Async::NotReady),
+                Ok(Async::Ready(None)) => return Ok(Async::Ready(())),
+                Ok(Async::Ready(Some((result, seqno)))) => {
+                    trace!(
+                        "produced message {:?} from {} at sequence id {}",
+                        result.message_id,
+                        result.producer_id,
+                        result.sequence_id
+                    );
+                    self.pending_acks.insert(seqno);
+                    let mut num_to_ack = 0;
+                    while self.pending_acks.remove(&self.seq_tail) {
+                        num_to_ack += 1;
+                        self.seq_tail += 1;
+                    }
+                    self.acker.ack(num_to_ack);
+                }
+                Err(e) => error!("future cancelled: {}", e),
```

Comment:
This could use a better message. The `kafka` error at least referenced that it was a "delivery" future. The term "future" probably has little meaning to the average Vector user.